File: src/shared/sftp-url.ts
export interface ParsedSftpUrl {
  username?: string
  password?: string
  host: string
  port?: number
  path: string
  scheme: 'sftp' | 'ssh' | 'scp'
}

export function looksLikeSftpInput(raw: string): boolean {
  const text = raw.trim()
  if (!text) return false
  if (/^(sftp|scp|ssh):\/\//i.test(text)) return true
  if (/^[a-zA-Z]:[\\/]/.test(text) || text.startsWith('\\\\') || text.startsWith('/')) return false
  return /^[^\s@/\\]+@[^@\s/\\]+/.test(text)
}

export function parseSftpUrl(raw: string): ParsedSftpUrl | null {
  const text = raw.trim()
  if (!text) return null

  const withScheme = /^(?:sftp|scp|ssh):\/\/(.+)$/i.exec(text)
  const scheme = (/^(sftp|scp|ssh):\/\//i.exec(text)?.[1]?.toLowerCase() || 'sftp') as ParsedSftpUrl['scheme']
  const rest = withScheme ? withScheme[1] : text
  if (!withScheme && !looksLikeSftpInput(text)) return null

  let username = ''
  let password: string | undefined
  let hostportpath = rest

  const at = rest.lastIndexOf('@')
  if (at >= 0) {
    const userinfo = rest.slice(0, at)
    hostportpath = rest.slice(at + 1)
    const colon = userinfo.indexOf(':')
    if (colon >= 0) {
      username = decode(userinfo.slice(0, colon))
      password = decode(userinfo.slice(colon + 1))
    } else {
      username = decode(userinfo)
    }
  }

  let host = hostportpath
  let port: number | undefined
  let path = '/'

  const slash = hostportpath.indexOf('/')
  let hostport = hostportpath
  if (slash >= 0) {
    hostport = hostportpath.slice(0, slash)
    path = hostportpath.slice(slash) || '/'
  } else {
    const scp = /^([^:]+):([^/].*)$/.exec(hostportpath)
    if (scp && !/^\d+$/.test(scp[2])) {
      hostport = scp[1]
      path = scp[2].startsWith('/') ? scp[2] : `/${scp[2]}`
    }
  }

  const portMatch = /^(.+):(\d+)$/.exec(hostport)
  if (portMatch) {
    host = portMatch[1]
    port = Number(portMatch[2]) || 22
  } else {
    host = hostport
  }

  host = host.replace(/^\[|\]$/g, '').trim()
  if (!host) return null
  if (!path.startsWith('/')) path = `/${path}`
  if (path.length > 1 && path.endsWith('/')) path = path.replace(/\/+$/, '') || '/'

  return { username: username || undefined, password, host, port, path, scheme }
}

export function formatSftpUrl(p: {
  username?: string
  host: string
  port?: number
  path: string
  scheme?: 'sftp' | 'ssh' | 'scp'
}): string {
  const scheme = p.scheme === 'ssh' ? 'ssh' : 'sftp'
  const user = p.username ? `${p.username}@` : ''
  const port = p.port && p.port !== 22 ? `:${p.port}` : ''
  const path = p.path || '/'
  return `${scheme}://${user}${p.host}${port}${path}`
}

function decode(value: string): string {
  try {
    return decodeURIComponent(value)
  } catch {
    return value
  }
}

File: src/CompareShow/Core/SftpUrl.cs
using System.Text.RegularExpressions;

namespace CompareShow.Core;

public sealed class ParsedSftpUrl
{
    public string? Username { get; set; }
    public string? Password { get; set; }
    public string Host { get; set; } = "";
    public int? Port { get; set; }
    public string Path { get; set; } = "/";
    public string Scheme { get; set; } = "sftp";
}

public static class SftpUrl
{
    public static bool LooksLikeSftpInput(string raw)
    {
        var text = raw.Trim();
        if (text.Length == 0) return false;
        if (Regex.IsMatch(text, @"^(sftp|scp|ssh)://", RegexOptions.IgnoreCase)) return true;
        if (Regex.IsMatch(text, @"^[a-zA-Z]:[\\/]") || text.StartsWith(@"\\") || text.StartsWith('/')) return false;
        return Regex.IsMatch(text, @"^[^\s@/\\]+@[^@\s/\\]+");
    }

    public static ParsedSftpUrl? Parse(string raw)
    {
        var text = raw.Trim();
        if (text.Length == 0) return null;

        var schemeMatch = Regex.Match(text, @"^(sftp|scp|ssh)://", RegexOptions.IgnoreCase);
        var scheme = schemeMatch.Success ? schemeMatch.Groups[1].Value.ToLowerInvariant() : "sftp";
        var rest = schemeMatch.Success ? text[schemeMatch.Length..] : text;
        if (!schemeMatch.Success && !LooksLikeSftpInput(text)) return null;

        string username = "";
        string? password = null;
        var hostportpath = rest;

        var at = rest.LastIndexOf('@');
        if (at >= 0)
        {
            var userinfo = rest[..at];
            hostportpath = rest[(at + 1)..];
            var colon = userinfo.IndexOf(':');
            if (colon >= 0)
            {
                username = Decode(userinfo[..colon]);
                password = Decode(userinfo[(colon + 1)..]);
            }
            else username = Decode(userinfo);
        }

        var path = "/";
        var hostport = hostportpath;
        var slash = hostportpath.IndexOf('/');
        if (slash >= 0)
        {
            hostport = hostportpath[..slash];
            path = hostportpath[slash..];
            if (path.Length == 0) path = "/";
        }
        else
        {
            var scp = Regex.Match(hostportpath, @"^([^:]+):([^/].*)$");
            if (scp.Success && !Regex.IsMatch(scp.Groups[2].Value, @"^\d+$"))
            {
                hostport = scp.Groups[1].Value;
                var p = scp.Groups[2].Value;
                path = p.StartsWith('/') ? p : "/" + p;
            }
        }

        string host;
        int? port = null;
        var portMatch = Regex.Match(hostport, @"^(.+):(\d+)$");
        if (portMatch.Success)
        {
            host = portMatch.Groups[1].Value;
            port = int.TryParse(portMatch.Groups[2].Value, out var n) ? n : 22;
        }
        else host = hostport;

        host = host.Trim().TrimStart('[').TrimEnd(']').Trim();
        if (host.Length == 0) return null;
        if (!path.StartsWith('/')) path = "/" + path;
        if (path.Length > 1) path = path.TrimEnd('/');
        if (path.Length == 0) path = "/";

        return new ParsedSftpUrl
        {
            Username = string.IsNullOrEmpty(username) ? null : username,
            Password = password,
            Host = host,
            Port = port,
            Path = path,
            Scheme = scheme
        };
    }

    public static string Format(string? username, string host, int? port, string path, string? scheme = null)
    {
        var sch = scheme == "ssh" ? "ssh" : "sftp";
        var user = string.IsNullOrEmpty(username) ? "" : username + "@";
        var p = port is > 0 and not 22 ? $":{port}" : "";
        return $"{sch}://{user}{host}{p}{path}";
    }

    static string Decode(string value)
    {
        try { return Uri.UnescapeDataString(value); }
        catch { return value; }
    }
}
